STACK-UP-3/warriorz-backend
# #171784088 Setup the API for User Role Settings

#### What does this PR do?
This PR implements the unit tests and logic for User Role/Permission Settings API. The app needs to support different user roles, with each role allowed a different level of access.

#### Description of Task to be completed?
- Setup unit tests for the API endpoints
- Setup database migrations and seeders
- Setup database models
- Setup controllers
- Setup route middleware
- Set the default role for a new User
- Add the User 'role' property to access token
- Verify that a role name must be unique

#### How should this be manually tested?
> First, you must be signed in as a User with the *Super Administrator* role.
> The `:id` refers to the unique id of a single resource in the database.

- Create a new Role in the database by sending an HTTP POST request to the API endpoint `/api/v1/roles`, with the request body as JSON containing the role properties 'name' and 'permissions'.
- Read all Roles from the database by sending an HTTP GET request to the API endpoint `/api/v1/roles`.
- Read a single Role from the database by sending an HTTP GET request to the API endpoint `/api/v1/roles/:id` (where `:id` is the unique id of the Role you want to retrieve).
- Update an existing Role in the database by sending an HTTP PATCH request to the API endpoint `/api/v1/roles/:id` (where `:id` is the unique id of the Role you want to update), with the request body as JSON containing the updated role properties 'name' and 'permissions'.
- Delete an existing Role in the database by sending an HTTP DELETE request to the API endpoint `/api/v1/roles/:id` (where `:id` is the unique id of the Role you want to delete).

As a *Super Administrator*, you are also able to do the following:

- Read all Users from the database by sending an HTTP GET request to the API endpoint `/api/v1/users`.
- Update an existing User's role in the database by sending an HTTP PATCH request to the API endpoint `/api/v1/users/:id` (where `:id` is the unique id of the User you want to update), with the request body as JSON containing the updated property 'role'.

[See screenshots below for more details...]

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[171784088](https://www.pivotaltracker.com/story/show/171784088)

#### Screenshots

##### Sign In as Super Administrator
![Sign In as Super Administrator](https://user-images.githubusercontent.com/1271441/79638919-94c9a600-8188-11ea-8563-2449f592fddd.png)

##### Show all existing Roles
![Show all existing Roles](https://user-images.githubusercontent.com/1271441/79638928-9dba7780-8188-11ea-9218-541c725bfd50.png)

##### Show a single Role
![Show a single Role](https://user-images.githubusercontent.com/1271441/79638932-a317c200-8188-11ea-9035-cb4c736a3806.png)

##### Create a new Role
![Create a new Role](https://user-images.githubusercontent.com/1271441/79638934-a8750c80-8188-11ea-8c02-3c09930cfcdd.png)

##### Show all Roles including newly-created Role
![Show all Roles including newly-created Role](https://user-images.githubusercontent.com/1271441/79638937-ae6aed80-8188-11ea-9ee7-1ce32df7cbde.png)

##### Update an existing Role
![Update an existing Role](https://user-images.githubusercontent.com/1271441/79639863-8aaaa600-818e-11ea-9fde-76ace80b149e.png)

##### Delete an existing Role
![Delete an existing Role](https://user-images.githubusercontent.com/1271441/79638940-b62a9200-8188-11ea-9f1a-740eaf615945.png)

##### Show all existing Users
![Show all existing Users](https://user-images.githubusercontent.com/1271441/79639867-926a4a80-818e-11ea-8503-89319ac86f90.png)

##### Update a User's role
![Update a User's role](https://user-images.githubusercontent.com/1271441/79639872-97c79500-818e-11ea-961f-750e0ff7f29a.png)
